crates/autopilot/src/domain/settlement/transaction.rs

```diff
@@ -1,18 +1,54 @@
-use crate::domain::eth;
+use {
+    crate::domain::{self, eth},
+    anyhow::anyhow,
+};
 
-/// An on-chain transaction that settled a solution.
-#[derive(Debug)]
+/// An on-chain transaction that settled a solution, with calldata in a valid
+/// format.
+#[derive(Debug, Clone)]
 pub struct Transaction {
     /// The hash of the transaction.
     pub hash: eth::TxId,
+    /// The associated auction id.
+    pub auction_id: domain::auction::Id,
     /// The address of the solver that submitted the transaction.
     pub solver: eth::Address,
-    /// The call data of the transaction.
-    pub input: eth::Calldata,
     /// The block number of the block that contains the transaction.
     pub block: eth::BlockNo,
     /// The gas used by the transaction.
     pub gas: eth::Gas,
     /// The effective gas price of the transaction.
     pub effective_gas_price: eth::EffectiveGasPrice,
+    /// The solution that was settled.
+    pub solution: domain::settlement::Solution,
 }
+
+impl Transaction {
+    pub fn new(
+        transaction: &eth::Transaction,
+        domain_separator: &eth::DomainSeparator,
+    ) -> anyhow::Result<Self> {
+        let (_, metadata) = transaction
+            .input
+            .0
+            .split_at(transaction.input.0.len() - META_DATA_LEN);
```

Comment:
Splitting the parsing of the regular calldata and the auction id in 2 separate steps is not able to detect the case when a solver did not append the auction id.
If you first decode all the bytes of a tx without the extra data this will succeed. And if you later take the last 8 bytes it will of course also work (because a tx than can be decoded will have more than 8 bytes).
But if you were to first cut off the 8 bytes and then decode the remaining bytes the operation would rightfully fail.

Reply:
Good one. I forwarded only the first part to solution constructor which should be good.